src/twisted/trial/test/test_tests.py

```diff
@@ -962,6 +962,21 @@ def test_multipleErrorsReported(self):
         self.assertEqual(error1.getErrorMessage(), "bar")
         self.assertEqual(error2.getErrorMessage(), "foo")
 
+    def test_cleanupRunsOnce(self):
+        """
+        A function registered as a cleanup is run once.
+        """
+        cleanups = []
+        self.test.addCleanup(lambda: cleanups.append(stage))
+        # It should get run this time.
+        stage = "first"
+        self.test.run(self.result)
+        # It should not get run the next time since it has not been
+        # re-registered.
+        stage = "second"
+        self.test.run(self.result)
+        self.assertEqual(cleanups, ["first"])
```

Comment:
This test is a just a little confusing because it took me a moment to realize that `addCleanup` is expected to be invoked *during* `run`, and not usually like this.  A more realistic test would be something that adds a cleanup each time *during* the synthetic test run, but asserts that the appropriate cleanup call is paired with each run and the streams don't get crossed.

This test would also pass if, for example, all cleanups just didn't run after the first test run.  But this is a pretty minor critique so I would not call it a blocker; if you want to clean up the test just a little before landing, go for it

Reply:
I agree that it's weird.  It seems to be the prevailing idiom for this test suite, though, so I'm inclined to go with consistency for now.  I would approve of a PR that rewrites all of `AddCleanupMixin` to use more realistic test cases (maybe I'll even get to it after some of the other functional fixes I'm working on).